### ISSUE TITLE
[ASK] question about cornac with implicit dataset

### ISSUE THREAD

**Issue by joy-more** [PreferredAI/cornac#254]
### Description

Hi,

When I used cornac with implicit datset, some evaluations make me puzzled(the "MAE" and "RMSE" were always 0, is it right or a bug?). Note that the implicit dataset in my experiments only have the positive feedback(the ratings is 1) while the negative feedbacks have beed removed.

when the epochs=20, the results were shown as follow:
![图片](https://user-images.githubusercontent.com/56193723/66254571-9f9f8900-e7aa-11e9-98fd-61c02737a892.png)

when the epochs=40, the results were shown as follow:
![图片](https://user-images.githubusercontent.com/56193723/66254635-8f3bde00-e7ab-11e9-876d-b21113bcc27b.png)

when the epochs=60, the results were shown as follow:
![图片](https://user-images.githubusercontent.com/56193723/66254641-9531bf00-e7ab-11e9-9d4d-99c1fc1c4725.png)

the codes have been used were shown as follow:
![图片](https://user-images.githubusercontent.com/56193723/66254667-fe193700-e7ab-11e9-9e87-07143509c0bd.png)

Thanks for your reading and your excellent jobs in this tool!

**Comment by tqtg** [comment]
Hi @joy-more 
Do you mind sharing your data file `music_rating_finished.txt` with us to reproduce the issue?

One more thing, do you observe the same issue when trying with some built-in datasets (e.g., MovieLens 100k)? Does your data file have the format of `user \t item \t rating`?

**Comment by joy-more** [comment]
> 
> 
> Hi @joy-more
> Do you mind sharing your data file `music_rating_finished.txt` with us to reproduce the issue?
> 
> One more thing, do you observe the same issue when trying with some built-in datasets (e.g., MovieLens 100k)? Does your data file have the format of `user \t item \t rating`?

Hi, would you mind send me your email so that I can share my dataset with you? The format is follow the UIR.

**Comment by joy-more** [comment]
>  data file `music_rating_finished.txt`

Hi, I've send the dataset to you. Thanks for your help.

**Comment by tqtg** [comment]
> > Hi @joy-more
> > Do you mind sharing your data file `music_rating_finished.txt` with us to reproduce the issue?
> > One more thing, do you observe the same issue when trying with some built-in datasets (e.g., MovieLens 100k)? Does your data file have the format of `user \t item \t rating`?
> 
> Hi, would you mind send me your email so that I can share my dataset with you? The format is follow the UIR.

Please send the data file to my email address tuantq.vnu@gmail.com.
Thanks.

**Comment by tqtg** [comment]
Hi @joy-more ,

For your dataset (implicit feedback), it makes more sense to use ranking metrics (e.g., Recall, NDCG) to evaluate model performance rather than using rating metrics (e.g., MAE, RMSE).

In this particular case, the reason why MAE and RMSE equal to 0.0 is that predicted rating from all models will go through clipping. Specifically, the predicted rating will be clipped so that `min_rating` <= `predicted_rating` <=`max_rating`, where `min_rating` and `max_rating` are obtained from training data ([details](https://github.com/PreferredAI/cornac/blob/cf04c245e028d6490853ee2304c34df5dfb8bc01/cornac/models/recommender.py#L122)). For your data, `min_rating` = `max_rating` = 1.0, and the compared ground-truth values all equal to 1.0 as well. Thus, MAE = RMSE = 0.0 is expected.

Hope this helps!

**Comment by joy-more** [comment]
> 
> 
> Hi @joy-more ,
> 
> For your dataset (implicit feedback), it makes more sense to use ranking metrics (e.g., Recall, NDCG) to evaluate model performance rather than using rating metrics (e.g., MAE, RMSE).
> 
> In this particular case, the reason why MAE and RMSE equal to 0.0 is that predicted rating from all models will go through clipping. Specifically, the predicted rating will be clipped so that `min_rating` <= `predicted_rating` <=`max_rating`, where `min_rating` and `max_rating` are obtained from training data ([details](https://github.com/PreferredAI/cornac/blob/cf04c245e028d6490853ee2304c34df5dfb8bc01/cornac/models/recommender.py#L122)). For your data, `min_rating` = `max_rating` = 1.0, and the compared ground-truth values all equal to 1.0 as well. Thus, MAE = RMSE = 0.0 is expected.
> 
> Hope this helps!

Thanks a lot. And should I sample some negative feedbacks  to supplement the dataset that only has pisitive feedback?

**Comment by tqtg** [comment]
You only need to provide positive feedback. In fact, all missing feedback is treated as negative and the sampling is already supported depending on the need of each individual model.

**Comment by joy-more** [comment]
> 
> 
> You only need to provide positive feedback. In fact, all missing feedback is treated as negative and the sampling is already supported depending on the need of each individual model.
“the sampling is already supported depending on the need of each individual model” is said that models of the Cornac have already implemented sampling operation?

**Comment by tqtg** [comment]
We’ve implemented some sampling-based iterator APIs to support model training in Cornac. You can find more details in [Dataset](https://cornac.readthedocs.io/en/latest/data.html#module-cornac.data.dataset) class.

**Comment by joy-more** [comment]
> 
> 
> We’ve implemented some sampling-based iterator APIs to support model training in Cornac. You can find more details in [Dataset](https://cornac.readthedocs.io/en/latest/data.html#module-cornac.data.dataset) class.

In this API: class cornac.data.dataset.Dataset(num_users, num_items, uid_map, iid_map, uir_tuple, timestamps=None, seed=None), How can I use it to sample dataset with any different samples? Just to chage the "seed" parameter?
